core/src/main/java/io/grpc/SynchronizationContext.java

```diff
@@ -80,29 +77,24 @@ public SynchronizationContext(UncaughtExceptionHandler uncaughtExceptionHandler)
    * have been or will eventually be run, while not requiring any more calls to {@code drain()}.
    */
   public final void drain() {
-    boolean drainLeaseAcquired = false;
-    while (true) {
-      Runnable runnable;
-      synchronized (lock) {
-        if (!drainLeaseAcquired) {
-          if (drainingThread != null) {
-            return;
-          }
-          drainingThread = Thread.currentThread();
-          drainLeaseAcquired = true;
-        }
-        runnable = queue.poll();
-        if (runnable == null) {
-          drainingThread = null;
-          break;
-        }
+    do {
+      if (!drainingThread.compareAndSet(null, Thread.currentThread())) {
+        return;
       }
       try {
-        runnable.run();
-      } catch (Throwable t) {
-        uncaughtExceptionHandler.uncaughtException(Thread.currentThread(), t);
+        Runnable runnable;
+        while ((runnable = queue.poll()) != null) {
+          try {
+            runnable.run();
+          } catch (Throwable t) {
+            uncaughtExceptionHandler.uncaughtException(Thread.currentThread(), t);
+          }
+        }
+      } finally {
+        drainingThread.set(null);
```

Comment:
There seems to be a race:

T1: call drain(): set drainingThread. Finish all tasks, exit loop
T2: call executeLater(): add a new task
T2: call drain(): try to set drainingThread, but is still set, give up
T1: still in drain(): reset drainingThread, then exit

End result: the task added by T2 will not be run unless someone calls drain() again.

Reply:
Nm, I see you have another loop to catch that.

Reply:
>T1: still in drain(): reset drainingThread, then exit

For T1, because there is  `while (!queue.isEmpty())` in the end, it will drain again, not exit.

Reply:
I agree without lock it's harder to prove the correctness and easier to break the correctness in the future though. But this seems to have the same complexity as the `SerializingExecutor`.